test/rockpaperscissors.js

```diff
@@ -0,0 +1,717 @@
+
+        const playerOneOwed = await rcp.balances(playerOne);
+        const playerTwoOwed = await rcp.balances(playerTwo);
+
+        assert.strictEqual(playerOneOwed.toString(10), "20");
+        assert.strictEqual(playerTwoOwed.toString(10), "0");
+
+    });
+
+    it("Player two can stake their winnings in a subsequent game", async () => {
+        await rcp.movePlayerOne(gameKey, toBN(10), {from: playerOne, value: toBN(10)});
+        await rcp.movePlayerTwo(gameKey, playerTwoMoveHash, {from: playerTwo, value: toBN(10)});
+        await rcp.revealPlayerOne(playerOneSecretBytes32, ROCK, {from: playerOne});
+        await rcp.revealPlayerTwo(gameKey, playerTwoSecretBytes32, PAPER, {from: playerTwo});
+
+        const initPlayerTwoOwed = await rcp.balances(playerTwo);
+        assert.strictEqual(initPlayerTwoOwed.toString(10), "19");
+
+        const playerMoveHash = await rcp.createPlayerOneMoveHash(playerOneSecretBytes32, ROCK, {from: playerTwo});
+        await rcp.movePlayerOne(playerMoveHash, toBN(10), {from: playerTwo});
+
+        const playerTwoOwed = await rcp.balances(playerTwo);
+        assert.strictEqual(playerTwoOwed.toString(10), "9");
+
+    });
+
+    it("Player ends the game early, collects the stake and withdraws their balance", async () => {
+
+        await rcp.movePlayerOne(gameKey, toBN(10), {from: playerOne, value: toBN(10)});
+
+        const txObj = await rcp.playerOneEndsGame(gameKey, {from: playerOne});
+
+        const initPlayerOneEthBalance = toBN(await getBalance(playerOne));
+
+        await truffleAssert.eventEmitted(txObj, 'PlayerOneEndsGame', (ev) => {
+
+            return  ev.player === playerOne &&
+                    ev.gameKey === gameKey &&
+                    ev.amount.toString(10) === "10"
+        },'PlayerOneEndsGame event is emitted');
+
+        const playerOneOwed = await rcp.balances(playerOne);
+        assert.strictEqual(playerOneOwed.toString(10), "10");
+
+        // Recipient withdraws
+        const txObj2 = await rcp.withdraw(toBN(5), {from: playerOne});
+
+        await truffleAssert.eventEmitted(txObj2, 'WithDraw', (ev) => {
+
+            return  ev.player === playerOne &&
+                    ev.amount.toString(10) === "5"
+        },'WithDraw event is emitted');
+
+        const cost = await getGasCost(txObj2);
+
+        // Get the player one's ETH and contract balances
+        const playerOneEthBalance = toBN(await getBalance(playerOne));
+        const playerOneOwed2 = toBN(await rcp.balances(playerOne));
+        const expectedPlayerOneEthBalance = initPlayerOneEthBalance.add(toBN(5)).sub(cost).toString(10);
+
+        assert.strictEqual(playerOneEthBalance.toString(10), expectedPlayerOneEthBalance);
+        assert.strictEqual(playerOneOwed2.toString(10), "5");
+
+    });
+
+
+    it("In Paper vs Rock, Paper wins", async () => {
+        const isPaperWinner = await rcp.isWinner(PAPER, ROCK);
+
+        assert.isTrue(isPaperWinner);
```

Comment:
```javascript
assert.isTrue(await rcp.isWinner(PAPER, ROCK));
```

Reply:
I am tempted to say that you could put them all in 1 `it`, and also to check the `isFalse` on the other combinations, right?
Plus, what happens when you pass invalid data?

Reply:
OK. Yes. I also considered one test block.

Reply:
> Plus, what happens when you pass invalid data?

I have added back this check:

` require(playerTwoMove < 3, INVALID_MOVE_MSG);`
 
 When the moves are passed in revealPlayerOne() and in revealPlayerTwo().
 
 The values should hit these require()s before they are passed into isWinner();

Reply:
I can see that if it was intended for public use then I would also add `require()`s  in `isWinner()`.

Reply:
> `require(playerTwoMove < 3`

I really think that you should Rock at `1`, not `0`.

> add `require()`s in `isWinner()`

It is not a good idea a require in there.

Your `isWinner` function could instead return an enum. DRAW, LEFTWIN, RIGHTWIN, INCORRECT.
That would be more generic, and avoid the ambiguity of returning `false`: does it mean left wins? Or is it a draw?
Right now, you handle a draw before calling this `isWinner` function. This is something that is not apparent and so will trip other people using `isWinner` as is, and introduce bugs in the future.